cmd/checkheader/main.go

```diff
@@ -0,0 +1,267 @@
+
+const description = `
+Example: A check operation:
+  go run ./cmd/checkheader --ossl-include /usr/local/src/openssl-1.1.1/include ./openssl/openssl_funcs.h 
+Checkheader generates a C program, the compilation of which verifies types and functions defined in the target
+header file match the definitions in --ossl-include.
+`
+
+var osslInclude = flag.String("ossl-include", "", "OpenSSL include directory. Required.")
+var work = flag.Bool("work", false, "print the name of the temporary C program file and do not delete it when exiting.")
+
+func main() {
+	flag.Usage = func() {
+		fmt.Fprintf(flag.CommandLine.Output(), "\nUsage:\n")
+		flag.PrintDefaults()
+		fmt.Fprintf(flag.CommandLine.Output(), "%s\n\n", description)
+	}
+	flag.Parse()
+	if flag.NArg() != 1 {
+		flag.Usage()
+		os.Exit(1)
+	}
+	if *osslInclude == "" {
+		fmt.Fprintln(flag.CommandLine.Output(), "required flag not provided: --ossl-include")
+		flag.Usage()
+		os.Exit(1)
+	}
+	if _, err := os.Stat(*osslInclude); err != nil {
+		log.Fatalf("OpenSSL include directory not found: %v\n", err)
+	}
+	s, err := generate(flag.Arg(0))
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = gccRun(s)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func gccRun(program string) error {
+	f, err := os.CreateTemp("", "go-crypto-openssl-*.c")
+	if err != nil {
+		log.Fatal(err)
+	}
+	name := f.Name()
+	if !*work {
+		defer os.Remove(name)
+	} else {
+		defer log.Println(name)
+	}
+	if _, err = f.WriteString(program); err != nil {
+		log.Fatal(err)
+	}
+	if err := f.Close(); err != nil {
+		log.Fatal(err)
+	}
+	// gcc will fail to compile the generated C header if
+	// any of the static checks fails. If it succeed it means
+	// the checked header matches the OpenSSL definitions.
+	p := exec.Command("gcc",
+		"-c",                           // skip linking
+		"-Werror",                      // promote all warnings to errors
+		"-Wno-deprecated-declarations", // deprecation warnings are expected
+		"-isystem", *osslInclude,       // OpenSSL include from --ossl-include must be prefered over system includes
+		"-o", "/dev/null", // discard output
+		name)
+	p.Stdout = os.Stdout
+	p.Stderr = os.Stderr
+	return p.Run()
+}
+
+func generate(header string) (string, error) {
+	f, err := os.Open(header)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	var b strings.Builder
+	sc := bufio.NewScanner(f)
+	var i int
+	for sc.Scan() {
+		l := strings.TrimSpace(sc.Text())
+		if tryConvertDirective(&b, l) {
+			continue
+		}
+		if tryConvertTypedef(&b, l) {
+			continue
+		}
+		if tryConvertDefineFunc(&b, l, i) {
+			i++
+		}
+	}
+	if err := sc.Err(); err != nil {
+		return "", err
+	}
+	return b.String(), nil
+}
+
+func tryConvertDirective(w io.Writer, l string) bool {
+	if strings.HasPrefix(l, "// #include ") ||
+		strings.HasPrefix(l, "// #if ") ||
+		strings.HasPrefix(l, "// #endif") {
```

Comment:
Why can't this be simplified to 

```suggestion
	if strings.HasPrefix(l, "// #") {
```

Reply:
Good catch, at the end we want to support any C directive, and all start with `#`.